File: src/main/java/io/github/colintimbarndt/chat_emotes/data/unicode/pattern/PatternSlot.java
package io.github.colintimbarndt.chat_emotes.data.unicode.pattern;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;

public final class PatternSlot implements Matcher {
    private final @NotNull Matcher @NotNull[] matchers;
    private final int width;
    public final @NotNull SlotRange range;
    public final @NotNull String delimiter;
    public final boolean captured;
    private final boolean named;

    public PatternSlot(
            @NotNull Matcher @NotNull[] matchers,
            @NotNull SlotRange range,
            @NotNull String delimiter,
            boolean captured
    ) {
        this.matchers = matchers;
        this.range = range;
        this.delimiter = delimiter;
        int width = 0;
        for (var m : matchers) width += m.width();
        this.width = width;
        this.captured = captured;
        this.named = Arrays.stream(matchers).allMatch(Matcher::hasName);
    }

    @Override
    public MatchResult matchWith(String s, int offset) {
        int idxOffset = 0;
        for (var m : matchers) {
            final var r = m.matchWith(s, offset);
            if (r != null) return r.withOffset(idxOffset);
            idxOffset += m.width();
        }
        return null;
    }

    @Override
    public int width() {
        return width;
    }

    @Override
    public boolean hasName() {
        return named;
    }
}

File: common/src/main/kotlin/io/github/colintimbarndt/chat_emotes/common/data/unicode/pattern/PatternSlot.kt
package io.github.colintimbarndt.chat_emotes.common.data.unicode.pattern

/**
 * @see NamedPatternSlot
 */
data class PatternSlot(
    private val matchers: Array<Matcher>,
    override val range: SlotQuantifier,
    override val delimiter: String,
    override val captured: Boolean
) : IPatternSlot, Matcher {

    override val width: Int = matchers.asSequence().sumOf(Matcher::width)

    override val hasName: Boolean = matchers.asSequence().all(Matcher::hasName)

    override fun matchWith(s: String, offset: Int): MatchResult? {
        var idxOffset = 0
        for (m in matchers) {
            val r = m.matchWith(s, offset)
            if (r != null) return r.offsetBy(idxOffset)
            idxOffset += m.width
        }
        return null
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as PatternSlot

        return equalsHelper(PatternSlot::matchers, other)
    }

    override fun hashCode(): Int {
        var result = matchers.contentHashCode()
        result = 31 * result + range.hashCode()
        result = 31 * result + delimiter.hashCode()
        result = 31 * result + captured.hashCode()
        result = 31 * result + width
        result = 31 * result + hasName.hashCode()
        return result
    }
}